plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

```diff
@@ -109,75 +122,111 @@ runImportCommand _state (ImportCommandParams edit) = do
-          -- for every import, if it's lacking a explicit list, generate a code lens
-          Just (MinimalImportsResult minImports, posMapping) -> do
-            commands <-
-              sequence
-                [ generateLens pId _uri edit
-                  | (imp, Just minImport) <- minImports,
-                    Just edit <- [mkExplicitEdit pred posMapping imp minImport]
-                ]
-            return $ Right $ InL $ catMaybes commands
-          _ ->
-            return $ Right $ InL []
-    | otherwise =
-      return $ Right $ InL []
-
--- | If there are any implicit imports, provide one code action to turn them all
+lensProvider :: Recorder (WithPriority Log) -> PluginMethodHandler IdeState 'Method_TextDocumentCodeLens
+lensProvider _  state _ CodeLensParams {_textDocument = TextDocumentIdentifier {_uri}}
+  = pluginResponse $ do
+    nfp <- getNormalizedFilePath _uri
+    mbMinImports <- liftIO $ runAction "MinimalImports" state $ use MinimalImports nfp
+    case mbMinImports of
+      Just (MinimalImportsResult{forLens}) -> do
+        let lens = [ generateLens _uri range int
+                    | (range, int) <- forLens]
+        pure $ InL lens
+      _ ->
+        pure $ InL []
+  where generateLens :: Uri  -> Range -> Int -> CodeLens
+        generateLens uri range int =
+          CodeLens { _data_ = Just $ A.toJSON $ ResolveOne uri int
+                   , _range = range
+                   , _command = Nothing }
+
+lensResolveProvider :: Recorder (WithPriority Log) -> PluginMethodHandler IdeState 'Method_CodeLensResolve
+lensResolveProvider _ ideState plId cl@(CodeLens {_data_ = Just data_@(A.fromJSON -> A.Success (ResolveOne uri uid))})
+  = pluginResponse $ do
+    nfp <- getNormalizedFilePath uri
+    (MinimalImportsResult{forResolve}) <-
+      handleMaybeM "Unable to run Minimal Imports"
+      $ liftIO
+      $ runAction "MinimalImports" ideState $ use MinimalImports nfp
+    target <- handleMaybe "Unable to resolve lens" $ forResolve IM.!? uid
+    let updatedCodeLens = cl & L.command ?~  mkCommand plId target
+    pure updatedCodeLens
+  where mkCommand ::  PluginId -> TextEdit -> Command
+        mkCommand pId TextEdit{_newText} =
+          let title = abbreviateImportTitle _newText
+              _arguments = Just [data_]
+          in mkLspCommand pId importCommandId title _arguments
+lensResolveProvider _  _ _ (CodeLens {_data_ = Just (A.fromJSON -> A.Success (ResolveAll _))}) = do
+   pure $ Left $ ResponseError (InR ErrorCodes_InvalidParams) "Unexpected argument for lens resolve handler: ResolveAll" Nothing
+lensResolveProvider _  _ _ (CodeLens {_data_ = Just (A.fromJSON @EIResolveData -> (A.Error (T.pack -> str)))}) =
+  pure $ Left $ ResponseError (InR ErrorCodes_ParseError) str Nothing
+lensResolveProvider _  _ _ (CodeLens {_data_ = _}) = do
+   pure $ Left $ ResponseError (InR ErrorCodes_InvalidParams) "Unexpected argument for lens resolve handler: (Probably Nothing)" Nothing
```

Comment:
I'd just shove the whole data `Value` in the error tbh. If this ever happens you can't have too much info